### PR TITLE
Functions: connection handler

### PR DESCRIPTION
**Pull request by pinebit**

### CODE REVIEW

**Review by bolekk — CHANGES_REQUESTED** [review]

File: core/services/functions/connector_handler.go
Comment (pinebit) [reply core/services/functions/connector_handler.go:113]: We don't need to set Success = false, because it is false by default.
Helpers make little sense, since I have the single response path. The flow seems clear - we construct the response (whatever it is success or failure) and send.

File: core/services/functions/connector_handler.go
Comment (pinebit) [reply core/services/functions/connector_handler.go:27]: These can be updated once you merge your PR.

File: core/services/functions/listener_test.go
Comment (pinebit) [reply core/services/functions/listener_test.go:57]: As per linter, you should be omitting the type for rvalue as long as the compiler can derive the type without ambiguity.

File: core/services/functions/connector_handler.go
Comment (pinebit) [reply core/services/functions/connector_handler.go:62]: Added Allowlist, but I was not sure about what config parts are needed, just verify what I've done.
For rate limiting I believe this should go to the connector.

File: core/services/functions/connector_handler.go
Comment (bolekk) [inline core/services/functions/connector_handler.go:174]: If we decide that Gateway should combine responses from multiple nodes then the signature night need to be included in the payload. We can discuss that separately.

File: core/services/ocr2/plugins/functions/config/config.go
Comment (bolekk) [inline core/services/ocr2/plugins/functions/config/config.go:21]: For the allowlist to work you need to run UpdatePeriodically() or UpdateFromContract() in a loop. Check out functions.handler.go.

**Review by bolekk — CHANGES_REQUESTED** [review]
one allowlist comment

File: core/services/ocr2/plugins/functions/config/config.go
Comment (pinebit) [reply core/services/ocr2/plugins/functions/config/config.go:21]: This is not good. İ would wish the service (Allowlist) would be self-contained. E.g. implementing start/stop and hide all this logic internally.

Comment (bolekk) [reply core/services/ocr2/plugins/functions/config/config.go:21]: Yeah, I should've done that. I will convert to Start()/Stop() and own goroutine but in the meantime you can just call UpdatePeriodically() in your handler's Start().

Comment (pinebit) [reply core/services/ocr2/plugins/functions/config/config.go]: I reworked the Allowlist.

File: core/services/functions/connector_handler.go
Comment (pinebit) [reply core/services/functions/connector_handler.go:181]: This sets `msg.Signature`, and it gives it back with `SendToGateway`. The rest depends on the connector/gateway logic, you can include it or not, right? I mean this is not the functions' connector_handler responsibility anymore.

Comment (bolekk) [reply core/services/functions/connector_handler.go:181]: The problem is what to do if you want to aggregate responses from multiple nodes with signatures. You could pack multiple copies of Body into payload but that's wasteful. Another option is to add an extra signature inside payload and later combine those. We can discuss separately.

File: core/services/gateway/connector/connector.go
Comment (bolekk) [inline core/services/gateway/connector/connector.go:43]: I'm not a huge fan of this. What if there's a product that chooses not to validate signatures but use auth keys instead?
In the Gateway handler we pass full Message objects so let's at least do that here too, for consistency. I'll leave it up to you if you prefer to still Validate() before even calling Handle.

File: core/services/gateway/config/config.go
Comment (bolekk) [inline core/services/gateway/config/config.go:22]: This doesn't seem to be relevant to every DON. What if there's a DON that works with multiple chains, like CCIP?  What if somebody wants to check data on multiple chains in their handler? In Functions we're using it specifically for the allowlist so it makes sense to put it in Handler's config.

**Review by bolekk — CHANGES_REQUESTED** [review]

File: core/services/functions/connector_handler.go
Comment (pinebit) [reply core/services/functions/connector_handler.go:181]: My understanding: this is the connector, that executes as a part of a single node. I don't get why this would ever implement any aggregation logic. This is one side.
The other side is this: `SendToGateway` requires `api.Message` param, therefore I must populate all the Message fields. Otherwise this design looks inconsistent. I am considering all Gateway IO as a transport layer, which has independent signatures, etc. Transport is not meant to aggregate anything.
If I am wrong, then I don't get the whole design.

File: core/services/gateway/connector/connector.go
Comment (pinebit) [reply core/services/gateway/connector/connector.go:43]: > but use auth keys instead?

As I mentioned in another comment above, `api.Message.Signature` is a transport layer signature. It has nothing to do with product-specific logic/needs. That's how I see this. Let's discuss if your vision is different.
I believe that `core/services/gateway/connector/connector.go` should hide all signing and signature validation begin a part of the transport layer. The product-specific handler can do anything else, but should not care about transport needs. We already provide `.Sign()` method that the transport shall use.

File: core/services/gateway/config/config.go
Comment (pinebit) [reply core/services/gateway/config/config.go:22]: Convinced. Then I think we can take the chain ID from the job spec. I want to avoid specifying another Chain ID just for Handler, because a simple copy-paste error will make all parts inconsistent.

Comment (pinebit) [reply core/services/gateway/config/config.go]: Updated.

File: core/services/functions/connector_handler.go
Comment (bolekk) [reply core/services/functions/connector_handler.go:181]: Yes, the connector doesn't aggregate anything, I was talking about what happens next in the Gateway handler. Adding the signature here is correct, I was just flagging that we haven't implemented the aggregation part yet (in the Gateway handler) and it might be convenient to include another signature in the payload that covers a smaller set of fields. Again, let's discuss this separately.

File: core/services/gateway/connector/connector.go
Comment (bolekk) [reply core/services/gateway/connector/connector.go:43]: I wouldn't necessarily consider it a transport-layer signature because the transport doesn't depend on it. The channel between gateway and node is already authenticated on both sides. User authentication is up to the product in my view. The only thing that the transport really needs is the don_id to route messages to correct handler and that's it. I wanted to make everything else flexible.

Maybe some future Gateway use case would not care about signatures but require an auth key in the payload? Maybe some other future case will open the handler up to any user and only rate-limit globally? Who knows.

We can decide to follow your suggestion and treat the signature as a required part of every message and that will be fine for now. I agree that it is a cleaner and more consistent approach but the cost is reduced flexibility.

**Review by bolekk — APPROVED** [review]

**Pull request merged**
